File: neural_network/ruby/matrix.rb
class Matrix
	include Enumerable

	attr_reader :width, :height

	def initialize(*args)
		@matrix = []

		if args.first.is_a?(Array)
			@width = args.first.count
			@height = 0

			args.each do |arg|
				raise "all rows must be the same size" if arg.count != @width
				@matrix << arg
				@height += 1
			end
		elsif args.first.is_a?(Fixnum)
			@height = args[0]
			@width = args[1]

			@matrix = Array.new(@height){ Array.new(@width) }
		end

		if block_given?
			@matrix = @matrix.map do |row|
				row.map do |el|
					yield
				end
			end
		end
	end

	def each
		i = 0; j = 0

		while i < @height
			while j < @width
				yield(self[i, j])
				j += 1
			end
			j = 0; i += 1
		end

		@matrix
	end

	def map
		i = 0; j = 0 
		result = Matrix.new(@height, @width)

		while i < @height
			while j < @width
				result[i, j] = yield(self[i, j])
				j += 1
			end
			j = 0; i += 1
		end

		result
	end

	def [](i, j)
		@matrix[i][j]
	end

	def []=(i, j, val)
		@matrix[i][j] = val
	end

	def to_s
		@matrix.each { |row| puts row.to_s }
	end

	def arithmetic(operator, arg)
		i = 0; j = 0
		result = Matrix.new(@height, @width)

		while i < @height
			while j < @width
				if arg.is_a?(Numeric)
					result[i, j] = self[i, j].send(operator, arg)
				elsif arg.is_a?(Matrix)
					result[i, j] = self[i, j].send(operator, arg[i, j])
				end
				j += 1
			end
			j = 0; i += 1
		end

		result
	end

	def +(arg)
		arithmetic(:+, arg)
	end

	def -(arg)
		arithmetic(:-, arg)
	end

	def *(arg)
		arithmetic(:*, arg)
	end

	def /(arg)
		arithmetic(:/, arg)
	end
	
	def dot(other)
		raise "incompatible matrices" unless self.width == other.height

		r_height = self.height
		r_width = other.width
		result = Matrix.new(r_height, r_width)

		i = 0; j = 0
		while i < r_height
			while j < r_width
				row = self.row(i)
				col = other.col(j)
				result[i, j] = sum_product(row, col)
				j += 1
			end
			j = 0; i += 1
		end

		result
	end

	def row(idx)
		@matrix[idx]
	end

	def col(idx)
		column = []
		
		i = 0
		while i < @height
			column << self[i, idx]
			i += 1
		end

		column
	end

	def sum_product(array1, array2)
		sum = 0
		array1.each_with_index { |el, i| sum += el * array2[i] }
		sum
	end

	def transpose
		result = Matrix.new(@width, @height)

		i = 0; j = 0
		
		while i < @height
			while j < @width
				result[j, i] = self[i, j]
				j += 1
			end
			j = 0; i += 1
		end

		result
	end

	def to_a
		result = []
		self.each{ |el| result << el }
		result
	end
end

File: neural_network/matrix.js
const Matrix = function() {
  this.matrix = [];

  if (arguments[0] instanceof Array) {
    this.width = arguments[0].length;
    this.height = 0;

    let i = 0;
    while(i < arguments.length) {
      this.matrix.push(arguments[i]);
      this.height++;
      i++;
    }
  } else if (typeof(arguments[0]) === 'number' ) {
    this.height = arguments[0];
    this.width = arguments[1];

    let i = 0;
    while (i < this.height) {
      this.matrix.push([]);
      i++;
    }
  }
};

Matrix.prototype.set = function(i, j, val) {
  this.matrix[i][j] = val;
};

Matrix.prototype.get = function(i, j) {
  return this.matrix[i][j];
};

Matrix.prototype.each = function(callback) {
  let i = 0; let j = 0;

  while (i < this.height) {
    while (j < this.width) {
      callback(this.matrix[i][j], i, j);
      j++;
    }
    j = 0; i++;
  }

  return this.matrix;
};

Matrix.prototype.map = function(callback) {
  let result = new Matrix(this.height, this.width);

  let i = 0; let j = 0;
  while (i < this.height) {
    while (j < this.width) {
      result.set(i, j, callback(this.matrix[i][j], i, j));
      j++;
    }
    j = 0; i++;
  }

  return result;
};

Matrix.prototype.row = function(idx) {
  return this.matrix[idx];
};

Matrix.prototype.col = function(idx) {
  let column = [];

  let i = 0;
  while (i < this.height) {
    column.push(this.matrix[i][idx]);
    i++;
  }

  return column;
};

Matrix.prototype.dot = function(other) {
  if (this.width !== other.height) {
    throw "Incompatible matrices.";
  }

  let resultHeight = this.height;
  let resultWidth = other.width;
  let result = new Matrix(resultHeight, resultWidth);

  let i = 0; let j = 0;
  while (i < resultHeight) {
    while (j < resultWidth) {
      let row = this.row(i);
      let col = other.col(j);
      result.set(i, j, Matrix.sumProduct(row, col));
      j++;
    }
    j = 0; i++;
  }

  return result;
};

Matrix.sumProduct = function(arr1, arr2) {
  let sum = 0;

  arr1.forEach( (val, idx) =>
    sum += (arr1[idx] * arr2[idx])
  );

  return sum;
};

Matrix.prototype.transpose = function() {
  let result = new Matrix(this.width, this.height);

  let i = 0; let j = 0;
  while (i < this.height) {
    while (j < this.width) {
      result.set(j, i, this.get(i, j));
      j++;
    }
    j = 0; i++;
  }

  return result;
};

Matrix.prototype.times = function(other) {
  let result = new Matrix(this.height, this.width);

  let i = 0; let j = 0;
  while (i < this.height) {
    while (j < this.width) {
      result.set(i, j, this.get(i, j) * other.get(i, j));
      j++;
    }
    j = 0; i++;
  }

  return result;
};

Matrix.prototype.toArray = function() {
  let arr = [];

  this.each( x => arr.push(x) );

  return arr;
};

module.exports = Matrix;
